### PR TITLE
diagnostics: make paths to external items more visible

### LINKED ISSUE #23224: Potential error in error (warning) message
I am too new at rust to know if this is really an issue or not.

I was trying to get the sample "guess" program to work. At one point I got this warning:
> warning: pattern binding `Less` is named the same as one of the variants of the type `core::cmp::Ordering` [E0170]

So I change my namespace to use '**core**' and I got this error:
> error: failed to resolve. Use of undeclared type or module `core::cmp::Ordering`

In the end, the solution was to use "**std**::cmp::Ordering::Less" - so the bug(?) is the warning message steering me to the wrong namespace.

### LINKED ISSUE #23355: Error message suggests using `core` when it ought to suggest using `std`
Program:
```rust
fn main() {
    Vec::from_iter();
}
```
Output:
```
fri.rs:2:5: 2:19 error: type `collections::vec::Vec<_>` does not implement any method in scope named `from_iter`
fri.rs:2     Vec::from_iter();
             ^~~~~~~~~~~~~~
fri.rs:2:19: 2:19 help: methods from traits can only be called if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
fri.rs:2:19: 2:19 help: candidate #1: use `core::iter::FromIterator`
error: aborting due to previous error
```
Note that it suggests `core::iter::FromIterator` when it could be suggesting `std::iter::FromIterator`. Should a user follow this advice, they would need to also link libcore via `extern crate core`, which (IIRC) means that they will be opting out of 1.0-stable thanks to core's intentional lack of stability.

I don't know if it's worth nominating, but this is definitely a backcompat hazard *for users*.

### LINKED ISSUE #26635: Compiler suggests wrong import `collections::borrow` instead of `std::borrow`
Example
```rust
type T = u32;

pub fn main() {
    let boxed : Option<Box<T>> = Some(Box::new(5));

    let reference : Option<&T> = match boxed {
        Some(ref x) => Some(x.borrow()),
        None => None
    };
}
```

[results in](play.rust-lang.org):
```
<anon>:7:31: 7:39 error: no method named `borrow` found for type `&Box<u32>` in the current scope
<anon>:7         Some(ref x) => Some(x.borrow()),
                                       ^~~~~~~~
<anon>:7:31: 7:39 help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
<anon>:7:31: 7:39 help: candidate #1: use `collections::borrow::Borrow`
error: aborting due to previous error
playpen: application terminated with error code 101
Compilation failed.
```

it would be correct if it suggested to use `std::borrow::Borrow`

### LINKED ISSUE #27165: Issue with suggestion for Clone
I wrote an implementation for Clone but forgot to import it. Here's the error I got initially:

```
src/server/service.rs:123:67: 123:74 error: no method named `clone` found for type `server::service::OpenTime` in the current scope
src/server/service.rs:123             s.open_times.insert(key.clone().to_string(), (*value).clone());
                                                                                            ^~~~~~~
note: in expansion of for loop expansion
src/server/service.rs:122:9: 124:10 note: expansion site
src/server/service.rs:123:67: 123:74 help: items from traits can only be used if the trait is implemented and in scope; the following trait defines an item `clone`, perhaps you need to implement it:
src/server/service.rs:123:67: 123:74 help: candidate #1: `core::clone::Clone`
```

After importing core, this came next:

```
src/server/service.rs:8:1: 8:19 error: use of unstable library feature 'core'
src/server/service.rs:8 extern crate core;
```

It's a bit misleading to suggest core and not the std library version of it.

### PR DESCRIPTION
**Pull request by jseyfried**
This PR changes the reported path for an external item so that it is visible from at least one local module (i.e. it does not use any inaccessible external modules) if possible. If the external item's crate was declared with an `extern crate`, the path is guarenteed to use the `extern crate`.

Fixes #23224, fixes #23355, fixes #26635, fixes #27165.

r? @nrc

### CODE REVIEW

**Comment by alexcrichton** [comment]
Holy cow I thought I'd never live to see the day this got implemented.

Thank you @jseyfried!

:cake:

**Comment by mitaa** [comment]
Nice!

FYI, with #32293 there is also a refactoring going on here which lands soon. (https://github.com/rust-lang/rust/pull/32293/commits/684c734e603ce69480f1f56a48971df9f592bfbb, https://github.com/rust-lang/rust/pull/32293/commits/767c7897bd6a22994204619ccdba762550abd0ed, ...)

File: src/librustc_metadata/cstore.rs [hunk 1]
```diff
@@ -244,6 +247,52 @@ impl CStore {
     {
         self.extern_mod_crate_map.borrow().get(&emod_id).cloned()
     }
+
+    pub fn visible_parent_map<'a>(&'a self) -> ::std::cell::RefMut<'a, DefIdMap<DefId>> {
```
Comment (arielb1) [inline src/librustc_metadata/cstore.rs]: Why the `RefMut`?

Comment (jseyfried) [inline src/librustc_metadata/cstore.rs]: The `RefCell` has to be borrowed mutably in this function and I didn't want to needlessly re-borrow it immutably before returning. This function is only used in `librustc_metadata` so I don't think it makes much of a difference.

Comment (nrc) [inline src/librustc_metadata/cstore.rs]: Could you document this function please?

File: src/librustc_metadata/cstore.rs [hunk 2]
```diff
@@ -244,6 +247,52 @@ impl CStore {
...
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert(parent);
+                        bfs_queue.push_back(child);
+                    }
+                }
+            };
+
+            let croot = DefId { krate: cnum, index: CRATE_DEF_INDEX };
+            for child in self.crate_top_level_items(cnum) { add_child(bfs_queue, child, croot); }
```
Comment (nrc) [inline src/librustc_metadata/cstore.rs]: nit: spread over 3 lines

File: src/librustc_metadata/cstore.rs [hunk 3]
```diff
@@ -244,6 +247,52 @@ impl CStore {
...
+            while let Some(def) = bfs_queue.pop_front() {
+                for child in self.item_children(def) { add_child(bfs_queue, child, def); }
```
Comment (nrc) [inline src/librustc_metadata/cstore.rs]: and this one

File: src/librustc_metadata/csearch.rs [hunk 4]
```diff
@@ -140,6 +140,31 @@ impl<'tcx> CrateStore<'tcx> for cstore::CStore {
         })
     }
 
+    /// If possible, this returns a global path to the external item `def` that is visible from
+    /// at least one local module. If the external item's crate was declared with an
+    /// `extern crate`, the path is guarenteed to use the `extern crate`.
+    fn visible_item_path(&self, def: DefId) -> Option<Vec<hir_map::PathElem>> {
```
Comment (nrc) [inline src/librustc_metadata/csearch.rs]: It might be simpler to return item_path() rather than None if there is no visible item path

**Comment by nrc** [comment]
r+ with the nits addressed

File: src/librustc_metadata/cstore.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_metadata/cstore.rs]: Will do

File: src/librustc_metadata/csearch.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc_metadata/csearch.rs]: Good point

**Comment by jseyfried** [comment]
@nrc thanks for the feedback! I addressed your comments in the most recent commit.

We should probably wait for #32293 to land before merging this since there will be non-trivial conflicts (cc @nikomatsakis).

**Comment by eddyb** [comment]
@jseyfried I think this can be r=nrc with `src/test/pretty/issue-4264.pp` updated to use the new paths (see travis failure).

**Comment by alexcrichton** [comment]
This ended up passing all tests on bors, not sure what happened here so I'm just gonna merge manually

**Pull request merged**